### ISSUE TITLE
AmbiguousForeignKeysError error visible on SQLAlchemy 1.4.20 but not visible on 1.3.x version

### ISSUE THREAD

**Issue by ghevge** [sqlalchemy/sqlalchemy#6729]
I'm trying to migrate to SQLAlchemy 1.4.20 from 1.3.x , but in the process I've stumbled over the error below.
Any idea what is going on? 

The exact same code (available below) is working ok with the 1.3.x version.

```
m-middleware-server    | [tornado.application] 2021-07-07 17:19:49,096 ERROR web:1789: Uncaught exception GET /app-config-promotions?get_promotions_pending_approval=true&namespace_id=2 (172.26.0.16)
m-middleware-server    | HTTPServerRequest(protocol='http', host='boltwebserver:8888', method='GET', uri='/app-config-promotions?get_promotions_pending_approval=true&namespace_id=2', version='HTTP/1.0', remote_ip='172.26.0.16')
m-middleware-server    | Traceback (most recent call last):
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/tornado/web.py", line 1704, in _execute
m-middleware-server    |     result = await result
m-middleware-server    |   File "/bolt-webserver/bolt/api/handler/app_config_promotions.py", line 369, in get
m-middleware-server    |     app_config_promotions_tuple = query_union.all()
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2709, in all
m-middleware-server    |     return self._iter().all()
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2844, in _iter
m-middleware-server    |     result = self.session.execute(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1689, in execute
m-middleware-server    |     result = conn._execute_20(statement, params or {}, execution_options)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1582, in _execute_20
m-middleware-server    |     return meth(self, args_10style, kwargs_10style, execution_options)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 323, in _execute_on_connection
m-middleware-server    |     return connection._execute_clauseelement(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1443, in _execute_clauseelement
m-middleware-server    |     compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 518, in _compile_w_cache
m-middleware-server    |     compiled_sql = self._compiler(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 553, in _compiler
m-middleware-server    |     return dialect.statement_compiler(dialect, self, **kw)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 766, in __init__
m-middleware-server    |     Compiled.__init__(self, dialect, statement, **kwargs)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 455, in __init__
m-middleware-server    |     self.string = self.process(self.statement, **compile_kwargs)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 490, in process
m-middleware-server    |     return obj._compiler_dispatch(self, **kwargs)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
m-middleware-server    |     return meth(self, **kw)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 3011, in visit_select
m-middleware-server    |     compile_state = select_stmt._compile_state_factory(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 500, in create_for_statement
m-middleware-server    |     return klass.create_for_statement(statement, compiler, **kw)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 656, in create_for_statement
m-middleware-server    |     self._setup_for_generate()
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 678, in _setup_for_generate
m-middleware-server    |     self._legacy_join(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 1534, in _legacy_join
m-middleware-server    |     self._join_left_to_right(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 1615, in _join_left_to_right
m-middleware-server    |     _ORMJoin(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/orm/util.py", line 1620, in __init__
m-middleware-server    |     expression.Join.__init__(self, left, right, onclause, isouter, full)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/selectable.py", line 1010, in __init__
m-middleware-server    |     self.onclause = self._match_primaries(self.left, self.right)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/selectable.py", line 1151, in _match_primaries
m-middleware-server    |     return self._join_condition(left, right, a_subset=left_right)
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/selectable.py", line 1184, in _join_condition
m-middleware-server    |     cls._joincond_trim_constraints(
m-middleware-server    |   File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/selectable.py", line 1305, in _joincond_trim_constraints
m-middleware-server    |     raise exc.AmbiguousForeignKeysError(
m-middleware-server    | sqlalchemy.exc.AmbiguousForeignKeysError: Can't determine join between 'anon_1' and 'applications'; tables have more than one foreign key constraint relationship between them. Please specify the 'onclause' of this join explicitly.
```

The code behind this error is the following:

```
app_roles_query = self.user.app_roles\
            .join(Application, ApplicationConfig, Promotion)\
            .with_entities(Namespace, Application, ApplicationConfig, Promotion)\
            .filter(ApplicationConfig.id == Promotion.app_config_id)\
            .filter(Application.id == ApplicationConfig.app_id)\
            .filter(Application.namespace_id == Namespace.id)

        namespace_query = self.user.namespaces\
            .join(Application, ApplicationConfig, Promotion)\
            .with_entities(Namespace, Application, ApplicationConfig, Promotion)\
            .filter(ApplicationConfig.id == Promotion.app_config_id)\
            .filter(Application.id == ApplicationConfig.app_id)\
            .filter(Application.namespace_id == Namespace.id)\

        query_union = namespace_query.union(app_roles_query)
        if filter_promotions_pending_approval:
            query_union = query_union\
                .filter(Promotion.pending_approval.is_(true()))\
                .join(PromotionEnvironmentRegion, PromotionEnvironmentRegion.promotion_id == Promotion.id)\
                .filter(PromotionEnvironmentRegion.is_overridden.is_(false()))

        app_config_promotions_tuple = query_union.all()
```

**Comment by zzzeek** [comment]
cant be sure but the issue involves those complex joins you are writing, those should be migrated to a more modern format, using ON clauses (I'm just making these up, but this is the idea):

```
    self.user.app_roles\
            .join(Application, User.app_id == Application.id)\
            .join(ApplicationConfig, Application.foo_id == ApplicationConfig.config_id)\
           .join(Promotion, Promotion.application_config_id == ApplicationConfig.config_id)\
```

updated forms on join() are documented at https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query%20join#sqlalchemy.orm.Query.join

the form you are using there currently is old enough that it was broken up until 1.4.6, where I fixed it in #6203, but it would not be surprising if it behaves a little differently in complex situations like what you have above.

**Comment by zzzeek** [comment]
this would need a complete [MCVE](http://stackoverflow.com/help/mcve) (succinct, runnable, shows the failure, no dependencies)  in order for me to be able to determine the behavioral change.  reopen if one may be provided.
